### PR TITLE
Utilize ExternalDNS as well as ExternalIP

### LINKED ISSUE #1921: Utilize ExternalDNS as well as ExternalIP
**Is your feature request related to a problem? Please describe.**

ExternalDNS is not referenced as the address of the game server.
https://github.com/googleforgames/agones/blob/v1.10.0/pkg/gameservers/gameservers.go#L42-L57

In a situation where the game server is located in an IPv4 network and the game client is located in an IPv6 network, if you are able to use ExternalDNS, NAT64 will do the translation correctly and you will be able to communicate.

Currently, only ExternalIP is referenced, so you will get raw IPv4 addresses, and you will have to translate them manually.

The most typical case of a game client deployed in an IPv6 network is an Apple/iOS device.
They require the game application to work properly even on iOS devices deployed in IPv6 networks.

**Describe the solution you'd like**

ExternalDNS is the preferred reference.
Or, provide an option to browse ExternalDNS preferentially.

**Describe alternatives you've considered**

The application can get information about the node directly and use it.
However, that would reduce the appeal of Agones by half.

**Additional context**

https://developer.apple.com/support/ipv6/

### PR DESCRIPTION
**Pull request by nanasi880**
**What type of PR is this?**
/kind feature

**What this PR does / Why we need it**:
In a situation where the game server is located in an IPv4 network and the game client is located in an IPv6 network, if you are able to use ExternalDNS, NAT64 will do the translation correctly and you will be able to communicate.

Currently, only ExternalIP is referenced, so you will get raw IPv4 addresses, and you will have to translate them manually.

The most typical case of a game client deployed in an IPv6 network is an Apple/iOS device.
They require the game application to work properly even on iOS devices deployed in IPv6 networks.

**Which issue(s) this PR fixes**:

Closes #1921

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by aLekSer** [comment]
@nanasi880 Thanks for you PR. I think this could be added as an Alpha feature first. Please take a look at this guideline:
https://agones.dev/site/docs/guides/feature-stages/
You can follow this PR as an example:
https://github.com/googleforgames/agones/pull/1642

File: pkg/gameservers/gameservers.go
Comment (aLekSer) [inline pkg/gameservers/gameservers.go:68]: Why this is a duplicate to lines 46-53?

**Review by aLekSer — CHANGES_REQUESTED** [review]
Good looking PR, but please clean up the code and get rid of duplicate code

File: pkg/gameservers/gameservers.go
Comment (nanasi880) [reply pkg/gameservers/gameservers.go:68]: The code is not duplicated.
Lines 46-53 are a search for External DNS.
This code is a search for Internal DNS.

Did you point out that you should search both External DNS and Internal DNS at once?

**Comment by nanasi880** [comment]
@aLekSer 
Thanks for the review.
I added the document and updated the build spec and Makefile.

File: pkg/gameservers/gameservers.go
Comment (aLekSer) [reply pkg/gameservers/gameservers.go:68]: Ok, now I see the difference. Note that there is a similar code down the file:
```
if a.Type == corev1.NodeInternalIP && net.ParseIP(a.Address) != nil {
```
So probably this code is redundant. WDYT? The only difference is `net.ParseIP` call.

File: pkg/gameservers/gameservers.go
Comment (markmandel) [inline pkg/gameservers/gameservers.go:81]: I would be tempted to keep the signature of `address(...)` the same and pull runtime.FeatureEnabled(runtime.NodeExternalDNS) into the function.

That way it's all encapsulated in one place - and also easier to adjust as we go from alpha :arrow_right: beta :arrow_right: to stable.

Also means less changes across the entire PR too.

File: pkg/gameservers/gameservers.go
Comment (markmandel) [inline pkg/gameservers/gameservers.go:45]:
```suggestion
externalDNS := runtime.FeatureEnabled(runtime.NodeExternalDNS)
```

File: pkg/gameservers/gameservers.go
Comment (markmandel) [inline pkg/gameservers/gameservers.go:44]:
```suggestion
func address(node *corev1.Node) (string, error) {
```

File: pkg/gameservers/gameservers_test.go
Comment (markmandel) [inline pkg/gameservers/gameservers_test.go:56]: This is a great first start, but rather than passing through, we can change the feature flag on the fly.

[here is a good example](https://github.com/googleforgames/agones/blob/master/pkg/apis/agones/v1/gameserver_test.go#L131) of where we change the feature flag on during testing, but also lock the test so that it doesn't interfere with parallel running tests.

File: pkg/util/runtime/features.go
Comment (markmandel) [inline pkg/util/runtime/features.go:48]: I like the name :+1:

File: site/content/en/docs/Guides/feature-stages.md
Comment (markmandel) [inline site/content/en/docs/Guides/feature-stages.md:34]: We'll need to wrap this new entry in a `feature` shortcode.

You an see an example here: https://agones.dev/site/docs/contribute/#within-a-page

Since 1.110.0 RC is tomorrow (I expect this won't land in this release, just because timing), this will probably need to target 1.12.0 - but we can adjust later if needed.

File: site/content/en/docs/Guides/utilize-node-externaldns.md
Comment (markmandel) [inline site/content/en/docs/Guides/utilize-node-externaldns.md:11]: Not sure this warrants it's own specific entry into the documentation Guide section :thinking: ?

It feels to me like an extension of this faq entry: https://agones.dev/site/docs/faq/#how-are-ip-addresses-allocated-to-gameservers

Extending that would probably work fine, and if we find that that section is getting too large on review, we should have a whole section around Address resolution, rather than a section just for this particular feature.

How does that sound?

**Review by markmandel — COMMENTED** [review]
This is an excellent start! Added some of my own review notes as well :+1:

**Comment by markmandel** [comment]
Looks like you got bit by what #1930 just fixed. Updating branch so we can re-test.

**Comment by markmandel** [comment]
Feature freeze is off - so we can return to this now :+1:

File: site/content/en/docs/Guides/utilize-node-externaldns.md
Comment (nanasi880) [reply site/content/en/docs/Guides/utilize-node-externaldns.md:11]: That's exactly right.
I'll try to add documentation to the items you suggested :)

File: pkg/gameservers/gameservers.go
Comment (nanasi880) [reply pkg/gameservers/gameservers.go:81]: I thought the function itself would be more valuable if it were pure, but on this scale, it sounded like my preference.
Since we are not particular about it, we will try to keep the signature of the address function.

File: pkg/gameservers/gameservers.go
Comment (nanasi880) [reply pkg/gameservers/gameservers.go:68]: To be honest, I was quite confused. If I were to implement it like a table-based test, it would look like this.

```
func address(node *corev1.Node) (string, error) {
	var targetTypes []corev1.NodeAddressType
	if runtime.FeatureEnabled(runtime.NodeExternalDNS) {
		targetTypes = []corev1.NodeAddressType{
			corev1.NodeExternalDNS, corev1.NodeExternalIP, corev1.NodeInternalDNS, corev1.NodeInternalIP,
		}
	} else {
		targetTypes = []corev1.NodeAddressType{corev1.NodeExternalIP, corev1.NodeInternalIP}
	}

	for _, addressType := range targetTypes {
		for _, a := range node.Status.Addresses {
			if a.Type != addressType {
				continue
			}
			if addressType == corev1.NodeExternalIP || addressType == corev1.NodeInternalIP {
				if net.ParseIP(a.Address) != nil {
					return a.Address, nil
				}
			} else {
				return a.Address, nil
			}
		}
	}

	return "", errors.Errorf("Could not find an address for Node: %s", node.ObjectMeta.Name)
}
```

After writing this far, I didn't think this code was smart. The current code may be silly and dumb, but it is terribly simple and easy to read.
If the amount of code increases further, it should be revised, but for the current amount, I thought the simplicity was more valuable.

File: pkg/gameservers/gameservers.go
Comment (markmandel) [reply pkg/gameservers/gameservers.go:81]: Thanks! It also makes it consistent with other feature flag implementations.

**Comment by nanasi880** [comment]
@markmandel 
The code and documentation have been revised according to the review results.

File: site/content/en/docs/Guides/feature-stages.md
Comment (markmandel) [inline site/content/en/docs/Guides/feature-stages.md:34]: Can we also wrap this in a feature shortcode please, so it won't show in current documentation when we merge it:
See: https://agones.dev/site/docs/contribute/#within-a-page

(Depending on formatting, may need to replicate the table with the additional property)

File: site/content/en/docs/FAQ/_index.md
Comment (markmandel) [inline site/content/en/docs/FAQ/_index.md:131]: Can we wrap this in a feature shortcode please, so it won't show in current documentation when we merge it:
See: https://agones.dev/site/docs/contribute/#within-a-page

**Review by markmandel — COMMENTED** [review]
This is good to go, except for the two small documentation fixes :+1: Nice work!

File: site/content/en/docs/Guides/feature-stages.md
Comment (markmandel) [reply site/content/en/docs/Guides/feature-stages.md:34]: To make this easier, just merged a PR that has this feature shortcode code in it - so if you rebase against master, you can take advantage:
(Also be good to cleanup the newly created conflicts too 😀)

https://github.com/googleforgames/agones/blame/master/site/content/en/docs/Guides/feature-stages.md#L27-L44

**Comment by nanasi880** [comment]
@markmandel 
Wrapped the document with a short code.
I also rebased the code against the latest master branch and cleanup the commits.

**Review by markmandel — APPROVED** [review]
LGTM!

**Comment by markmandel** [comment]
@aLekSer can we get an approval from you? I can't merge as you are still left at "requested changes", and we have RC tomorrow, and I'd like to get this in.

File: pkg/gameservers/gameservers.go
Comment (aLekSer) [reply pkg/gameservers/gameservers.go:70]: That's right, thanks

**Review by aLekSer — APPROVED** [review]

**Comment by aLekSer** [comment]
Node SDK test have failed:
```
includes/sdk.mk:69: recipe for target 'run-sdk-command-node' failed
```

**Comment by nanasi880** [comment]
I'm running this locally right now and it appears to be working fine... (I'll look into it some more)

**Comment by aLekSer** [comment]
This command works for me also.
`make run-sdk-command COMMAND=test SDK_FOLDER=node`
Seems debian packages was not available for Cloud Build environment. We need to run tests again.

**Comment by nanasi880** [comment]
Can you trigger the test again?

**Comment by nanasi880** [comment]
Rebased the code to the latest master branch.

**Pull request merged**
